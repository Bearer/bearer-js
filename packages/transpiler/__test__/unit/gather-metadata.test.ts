import * as path from 'path'
import * as ts from 'typescript'

import Metadata from '../../src/metadata'
import GatherMetadata from '../../src/transformers/gather-metadata'
const fixtures = path.join(__dirname, '..', '__fixtures__')

describe('GaterMetadata transformer', () => {
  describe('a simple root component', () => {
    it('has empty input and output', () => {
      const metadata = new Metadata()
      transpileFile('root-component.tsx', metadata)
      expect(metadata).toMatchSnapshot()
    })
  })

  describe('a simple root component', () => {
    it('has input and output matching props', () => {
      const metadata = new Metadata()
      transpileFile('root-component-with-api.tsx', metadata)
      expect(metadata).toMatchSnapshot()
    })
  })
})

function transpileFile(filename: string, metadata: Metadata) {
  const files = [path.join(fixtures, 'gather-metadata', filename)]
  const compilerHost = ts.createCompilerHost({})

  const program = ts.createProgram(
    files,
    { experimentalDecorators: true, outDir: path.join(__dirname, '../../.build-transformers') },
    compilerHost
  )
  program.emit(undefined, undefined, undefined, undefined, {
    before: [GatherMetadata({ metadata })]
  })
}
