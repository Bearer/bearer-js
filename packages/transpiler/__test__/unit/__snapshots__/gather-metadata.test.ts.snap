// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GaterMetadata transformer a simple root component has empty input and output 1`] = `
Metadata {
  "components": Array [
    Object {
      "classname": "FeatDisplayRootComponent",
      "fileName": "/Users/tanguyantoine/devel/bearer/bearer/packages/transpiler/__test__/__fixtures__/gather-metadata/root-component.tsx",
      "finalTagName": "feat-display",
      "group": "feat",
      "initialTagName": "feat-display",
      "inputs": Array [],
      "isRoot": true,
      "outputs": Array [],
    },
  ],
  "prefix": null,
  "registerComponent": [Function],
  "suffix": null,
}
`;

exports[`GaterMetadata transformer a simple root component has input and output matching props 1`] = `
Metadata {
  "components": Array [
    Object {
      "classname": "FeatDisplayRootComponent",
      "fileName": "/Users/tanguyantoine/devel/bearer/bearer/packages/transpiler/__test__/__fixtures__/gather-metadata/root-component-with-api.tsx",
      "finalTagName": "complex-feature-display",
      "group": "complex-feature",
      "initialTagName": "complex-feature-display",
      "inputs": Array [
        Object {
          "default": "'ok'",
          "name": "aStringProp",
          "type": "string",
        },
        Object {
          "default": "5",
          "name": "aNumberProp",
          "type": "number",
        },
      ],
      "isRoot": true,
      "outputs": Array [
        Object {
          "name": "bearer:1234-scenario-name:complex-feature:namedEvent",
          "payloadFormat": "unspecified",
        },
        Object {
          "name": "nonTypedEvent",
          "payloadFormat": "unspecified",
        },
        Object {
          "name": "typedEvent",
          "payloadFormat": Object {
            "type": "object",
          },
        },
        Object {
          "name": "typedEventWithType",
          "payloadFormat": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "properties": Object {
              "anEnum": Object {
                "enum": Array [
                  "ko",
                  "ok",
                ],
                "type": "string",
              },
              "something": Object {
                "properties": Object {
                  "name": Object {
                    "type": "string",
                  },
                  "status": Object {
                    "enum": Array [
                      "closed",
                      "open",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "name",
                  "status",
                ],
                "type": "object",
              },
              "somethingElse": Object {
                "type": "string",
              },
            },
            "required": Array [
              "anEnum",
              "something",
              "somethingElse",
            ],
            "type": "object",
          },
        },
        Object {
          "name": "typedEventWithTypeArray",
          "payloadFormat": Object {},
        },
        Object {
          "name": "typedEventWithTypeString",
          "payloadFormat": Object {
            "type": "string",
          },
        },
        Object {
          "name": "typedEventWithTypeNumber",
          "payloadFormat": Object {
            "type": "number",
          },
        },
        Object {
          "name": "typedEventWithTypeBoolean",
          "payloadFormat": Object {
            "type": "boolean",
          },
        },
      ],
    },
  ],
  "prefix": null,
  "registerComponent": [Function],
  "suffix": null,
}
`;
